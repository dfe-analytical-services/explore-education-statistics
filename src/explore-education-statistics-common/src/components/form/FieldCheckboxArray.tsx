import createErrorHelper from '@common/validation/createErrorHelper';
import {
  connect,
  FieldArray,
  FieldArrayRenderProps,
  FormikContext,
} from 'formik';
import get from 'lodash/get';
import React, { ReactNode } from 'react';

interface InjectedProps<FormValues> extends FieldArrayRenderProps {
  error: string;
  value: FormValues[keyof FormValues];
}

interface FieldCheckboxArrayProps<FormValues> {
  children: (props: InjectedProps<FormValues>) => ReactNode;
  error?: string;
  name: string;
  showError?: boolean;
  validateOnChange?: boolean;
}

const FieldCheckboxArray = <T extends {}>({
  children,
  error,
  name,
  formik,
  showError = true,
  validateOnChange = false,
}: FieldCheckboxArrayProps<T> & { formik: FormikContext<T> }) => {
  return (
    <FieldArray
      name={name}
      validateOnChange={validateOnChange || formik.validateOnChange}
    >
      {({ form, ...arrayHelpers }) => {
        const { getError } = createErrorHelper(form);

        let errorMessage = error || getError(name);

        if (!showError) {
          errorMessage = '';
        }

        const value = get(form.values, name);

        return children({
          ...arrayHelpers,
          error: errorMessage,
          form,
          name,
          value,
        });
      }}
    </FieldArray>
  );
};

export default connect<FieldCheckboxArrayProps<{}>>(FieldCheckboxArray);
