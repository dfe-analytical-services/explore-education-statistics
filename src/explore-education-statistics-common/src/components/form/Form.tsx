import ErrorSummary, {
  ErrorSummaryMessage,
} from '@common/components/ErrorSummary';
import { FormIdContextProvider } from '@common/components/form/contexts/FormIdContext';
import useMountedRef from '@common/hooks/useMountedRef';
import useToggle from '@common/hooks/useToggle';
import logger from '@common/services/logger';
import isErrorLike from '@common/utils/error/isErrorLike';
import createErrorHelper from '@common/validation/createErrorHelper';
import { useFormikContext } from 'formik';
import camelCase from 'lodash/camelCase';
import get from 'lodash/get';
import isEqual from 'lodash/isEqual';
import React, {
  ReactNode,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';

interface Props {
  children: ReactNode;
  id: string;
  submitId?: string;
  showErrorSummary?: boolean;
  showSubmitError?: boolean;
}

/**
 * Form wrapper to integrate with Formik.
 *
 * This provides a bunch of conveniences for displaying errors
 * and linking to them correctly (in error message hrefs) as
 * long as certain conventions are followed.
 *
 * Fields with errors should have ids which share the form's
 * id and camelCased value key in Formik
 * e.g. if form id is `timePeriodForm`, then a Formik value with a
 * key of `startDate` should have a field with an id of
 * `timePeriodForm-startDate`.
 *
 * Additional errors upon submitting the form e.g. from server
 * requests will also be added to the error summary. These
 * will link to the `submitId` prop.
 */
const Form = ({
  children,
  id,
  submitId = `${id}-submit`,
  showErrorSummary = true,
  showSubmitError = false,
}: Props) => {
  const isMounted = useMountedRef();

  const formik = useFormikContext();
  const { errors, touched, values, submitCount, submitForm } = formik;

  const previousValues = useRef(values);

  const { getAllErrors } = createErrorHelper({
    errors,
    touched,
  });

  const [hasSummaryFocus, toggleSummaryFocus] = useToggle(false);
  const [submitError, setSubmitError] = useState<ErrorSummaryMessage>();

  useEffect(() => {
    if (!isMounted.current) {
      return;
    }

    // If form has changed at all, we should remove the submit error
    if (!submitCount || !isEqual(values, previousValues.current)) {
      setSubmitError(undefined);
    }

    previousValues.current = values;
  }, [submitError, submitCount, values, isMounted]);

  const allErrors = useMemo(() => {
    const summaryErrors: ErrorSummaryMessage[] = Object.entries(getAllErrors())
      .filter(([errorName]) => get(touched, errorName))
      .map(([errorName, message]) => ({
        id: `${id}-${camelCase(errorName)}`,
        message: typeof message === 'string' ? message : '',
      }));

    return submitError ? [...summaryErrors, submitError] : summaryErrors;
  }, [getAllErrors, id, submitError, touched]);

  const handleSubmit = useCallback(
    async event => {
      toggleSummaryFocus.off();
      setSubmitError(undefined);
      event.preventDefault();

      try {
        await submitForm();
      } catch (error) {
        if (showSubmitError && isErrorLike(error)) {
          logger.error(error);

          if (isMounted.current) {
            setSubmitError({
              id: submitId,
              message: error.message,
            });
          }
        } else {
          throw error;
        }
      } finally {
        if (isMounted.current) {
          toggleSummaryFocus.on();
        }
      }
    },
    [toggleSummaryFocus, submitForm, showSubmitError, isMounted, submitId],
  );

  return (
    <FormIdContextProvider id={id}>
      <form id={id} onSubmit={handleSubmit}>
        {showErrorSummary && (
          <ErrorSummary
            errors={allErrors}
            id={`${id}-summary`}
            focusOnError={hasSummaryFocus}
            onFocus={toggleSummaryFocus.off}
          />
        )}

        {children}
      </form>
    </FormIdContextProvider>
  );
};

export default Form;
