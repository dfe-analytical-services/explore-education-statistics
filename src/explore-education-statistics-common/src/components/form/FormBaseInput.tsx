import ErrorMessage from '@common/components/ErrorMessage';
import FormLabel, { FormLabelProps } from '@common/components/form/FormLabel';
import classNames from 'classnames';
import React, {
  ChangeEventHandler,
  FocusEventHandler,
  KeyboardEventHandler,
  memo,
  MouseEventHandler,
  ReactNode,
} from 'react';

export interface FormBaseInputProps extends FormLabelProps {
  addOn?: ReactNode;
  addOnContainerClassName?: string;
  className?: string;
  disabled?: boolean;
  error?: ReactNode | string;
  hint?: string;
  id: string;
  list?: string;
  name: string;
  width?: 20 | 10 | 5 | 4 | 3 | 2;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  onChange?: ChangeEventHandler<HTMLInputElement>;
  onClick?: MouseEventHandler<HTMLInputElement>;
  onKeyPress?: KeyboardEventHandler<HTMLInputElement>;
}

interface HiddenProps {
  defaultValue?: string | number;
  type?: 'text' | 'number' | 'color' | 'search';
  value?: string | number;
}

const FormBaseInput = ({
  addOn,
  addOnContainerClassName,
  className,
  error,
  hint,
  id,
  hideLabel,
  label,
  labelSize,
  width,
  type = 'text',
  ...props
}: FormBaseInputProps & HiddenProps) => {
  const input = (
    <input
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
      aria-describedby={
        classNames({
          [`${id}-error`]: !!error,
          [`${id}-hint`]: !!hint,
        }) || undefined
      }
      className={classNames('govuk-input', className, {
        [`govuk-input--width-${width}`]: width !== undefined,
        'govuk-input--error': !!error,
      })}
      id={id}
      type={type}
    />
  );

  return (
    <>
      <FormLabel
        id={id}
        label={label}
        labelSize={labelSize}
        hideLabel={hideLabel}
      />

      {hint && (
        <span id={`${id}-hint`} className="govuk-hint">
          {hint}
        </span>
      )}

      {error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}
      {addOn ? (
        <div className={classNames('dfe-flex', addOnContainerClassName)}>
          {input}
          {addOn}
        </div>
      ) : (
        input
      )}
    </>
  );
};

export default memo(FormBaseInput);
