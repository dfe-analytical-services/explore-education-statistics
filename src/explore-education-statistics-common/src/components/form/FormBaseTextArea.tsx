import ErrorMessage from '@common/components/ErrorMessage';
import FormLabel, { FormLabelProps } from '@common/components/form/FormLabel';
import classNames from 'classnames';
import React, {
  ChangeEventHandler,
  FocusEventHandler,
  KeyboardEventHandler,
  MouseEventHandler,
  ReactNode,
  Ref,
  memo,
} from 'react';

export interface FormTextAreaProps
  extends Pick<FormLabelProps, 'hideLabel' | 'label'> {
  className?: string;
  disabled?: boolean;
  error?: ReactNode | string;
  hint?: string;
  id: string;
  inputRef?: Ref<HTMLTextAreaElement>;
  maxLength?: number;
  name: string;
  rows?: number;
  value?: string;
  onBlur?: FocusEventHandler<HTMLTextAreaElement>;
  onChange?: ChangeEventHandler<HTMLTextAreaElement>;
  onClick?: MouseEventHandler<HTMLTextAreaElement>;
  onKeyPress?: KeyboardEventHandler<HTMLTextAreaElement>;
}

function FormBaseTextArea({
  className,
  disabled,
  error,
  hint,
  id,
  inputRef,
  hideLabel,
  label,
  maxLength,
  name,
  rows = 5,
  value,
  onBlur,
  onChange,
  onClick,
  onKeyPress,
}: FormTextAreaProps) {
  return (
    <>
      <FormLabel id={id} label={label} hideLabel={hideLabel} />

      {hint && (
        <div id={`${id}-hint`} className="govuk-hint">
          {hint}
        </div>
      )}

      {error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}

      <textarea
        aria-describedby={
          classNames({
            [`${id}-error`]: !!error,
            [`${id}-hint`]: !!hint,
            [`${id}-info`]: !!maxLength,
          }) || undefined
        }
        className={classNames('govuk-textarea', className, {
          'govuk-textarea--error': !!error,
        })}
        disabled={disabled}
        id={id}
        name={name}
        ref={inputRef}
        onBlur={onBlur}
        onChange={onChange}
        onClick={onClick}
        onKeyPress={onKeyPress}
        rows={rows}
        value={value}
      />
    </>
  );
}

export default memo(FormBaseTextArea);
