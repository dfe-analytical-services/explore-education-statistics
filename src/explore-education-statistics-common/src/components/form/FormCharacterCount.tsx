import classNames from 'classnames';
import React from 'react';

interface Props {
  id: string;
  maxLength: number;
  value?: string;
}
export default function FormCharacterCount({ id, maxLength, value }: Props) {
  const remaining = maxLength - (value?.trim().length ?? 0);

  return (
    <div
      aria-live="polite"
      className={classNames('govuk-character-count__message', {
        'govuk-hint': remaining >= 0,
        'govuk-error-message': remaining < 0,
      })}
      id={`${id}-info`}
    >
      {remaining >= 0
        ? `You have ${remaining} character${
            remaining !== 1 ? 's' : ''
          } remaining`
        : `You have ${Math.abs(remaining)} character${
            remaining !== -1 ? 's' : ''
          } too many`}
    </div>
  );
}
