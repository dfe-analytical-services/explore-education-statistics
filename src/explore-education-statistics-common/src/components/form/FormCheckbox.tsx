import classNames from 'classnames';
import React, { ChangeEvent, FocusEventHandler, memo, ReactNode } from 'react';

export type OtherCheckboxChangeProps = Pick<FormCheckboxProps, 'label'>;

export type CheckboxChangeEventHandler = (
  event: ChangeEvent<HTMLInputElement>,
  otherCheckboxProps: OtherCheckboxChangeProps,
) => void;

export interface FormCheckboxProps {
  checked?: boolean;
  defaultChecked?: boolean;
  className?: string;
  conditional?: ReactNode;
  id: string;
  hint?: string;
  label: string;
  boldLabel?: boolean;
  name: string;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  onChange?: CheckboxChangeEventHandler;
  value: string;
  disabled?: boolean;
}

const FormCheckbox = ({
  checked,
  defaultChecked,
  className,
  conditional,
  id,
  hint,
  label,
  boldLabel = false,
  name,
  onBlur,
  onChange,
  value,
  disabled = false,
}: FormCheckboxProps) => {
  return (
    <>
      <div className={classNames('govuk-checkboxes__item', className)}>
        <input
          aria-describedby={hint ? `${id}-item-hint` : undefined}
          className="govuk-checkboxes__input"
          checked={checked}
          defaultChecked={defaultChecked}
          id={id}
          name={name}
          onBlur={onBlur}
          onChange={event => {
            if (onChange) {
              onChange(event, { label });
            }
          }}
          type="checkbox"
          value={value}
          disabled={disabled}
        />
        <label className="govuk-label govuk-checkboxes__label" htmlFor={id}>
          {boldLabel ? <strong>{label}</strong> : label}
        </label>
        {hint && (
          <span
            id={`${id}-item-hint`}
            className="govuk-hint govuk-checkboxes__hint"
          >
            {hint}
          </span>
        )}
      </div>
      {conditional && (
        <div
          className={classNames('govuk-checkboxes__conditional', {
            'govuk-checkboxes__conditional--hidden': !checked,
          })}
        >
          {conditional}
        </div>
      )}
    </>
  );
};

export default memo(FormCheckbox);
