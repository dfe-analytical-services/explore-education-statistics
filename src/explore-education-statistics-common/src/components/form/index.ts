export { default as Form } from './Form';
export { default as FormCheckbox } from './FormCheckbox';
export { default as FormCheckboxGroup } from './FormCheckboxGroup';
export { default as FormCheckboxSearchSubGroups } from './FormCheckboxSearchSubGroups';
export { default as FormCheckboxSearchGroup } from './FormCheckboxSearchGroup';
export { default as FormFieldCheckbox } from './FormFieldCheckbox';
export { default as FormFieldCheckboxGroup } from './FormFieldCheckboxGroup';
export { default as FormFieldCheckboxSearchGroup } from './FormFieldCheckboxSearchGroup';
export { default as FormFieldCheckboxSearchSubGroups } from './FormFieldCheckboxSearchSubGroups';
export { default as FormFieldRadioGroup } from './FormFieldRadioGroup';
export { default as FormFieldSelect } from './FormFieldSelect';
export { default as FormFieldTextArea } from './FormFieldTextArea';
export { default as FormFieldTextInput } from './FormFieldTextInput';
export { default as FormFieldset } from './FormFieldset';
export { default as FormGroup } from './FormGroup';
export { default as FormRadio } from './FormRadio';
export { default as FormRadioGroup } from './FormRadioGroup';
export { default as FormSelect } from './FormSelect';
export { default as FormTextInput } from './FormTextInput';
export { default as FormTextSearchInput } from './FormTextSearchInput';
