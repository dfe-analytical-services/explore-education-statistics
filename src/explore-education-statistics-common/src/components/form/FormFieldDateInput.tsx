import { useFormIdContext } from '@common/components/form/contexts/FormIdContext';
import FormFieldset from '@common/components/form/FormFieldset';
import FormNumberInput from '@common/components/form/FormNumberInput';
import { FormGroup } from '@common/components/form/index';
import { PartialDate } from '@common/utils/date/partialDate';
import delay from '@common/utils/delay';
import parseNumber from '@common/utils/number/parseNumber';
import { isValid, parse } from 'date-fns';
import { FormikErrors, useField } from 'formik';
import last from 'lodash/last';
import React, { ChangeEvent, useCallback, useRef, useState } from 'react';

interface Props<FormValues> {
  hint?: string;
  id?: string;
  legend: string;
  legendSize?: 'xl' | 'l' | 'm' | 's';
  name: keyof FormValues | string;
  partialDateType?: 'dayMonthYear' | 'monthYear';
  showError?: boolean;
  type?: 'date' | 'partialDate';
}

function FormFieldDateInput<FormValues>({
  id: customId,
  name,
  showError = true,
  hint,
  legend,
  legendSize = 'l',
  partialDateType = 'dayMonthYear',
  type = 'date',
}: Props<FormValues>) {
  const { fieldId } = useFormIdContext();
  const id = fieldId(name as string, customId);

  const isFocused = useRef(false);

  const [field, meta, helpers] = useField<
    Date | Partial<PartialDate> | undefined
  >(name as string);

  const [values, setValues] = useState<Partial<PartialDate>>(() => {
    if (field.value instanceof Date) {
      return {
        day: field.value?.getDate(),
        month: field.value?.getMonth()
          ? // eslint-disable-next-line no-unsafe-optional-chaining
            field && field.value?.getMonth() + 1
          : undefined,
        year: field.value?.getFullYear(),
      };
    }

    return {
      day: field.value?.day,
      month: field.value?.month,
      year: field.value?.year,
    };
  });

  const error =
    showError && meta.error && meta.touched
      ? (meta.error as FormikErrors<PartialDate> | string)
      : {};

  const errorMessage =
    typeof error === 'string' ? error : Object.values(error)[0] ?? '';

  const handleChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      const key = last(event.target.name.split('.')) as keyof PartialDate;

      if (!['day', 'month', 'year'].includes(key)) {
        throw new Error(`Invalid key for day/month/year field: ${key}`);
      }

      const inputValue = parseNumber(event.target.value);

      const day = key === 'day' ? inputValue : values.day;
      const month = key === 'month' ? inputValue : values.month;
      const year = key === 'year' ? inputValue : values.year;

      if (type === 'date') {
        const date = parse(`${year}-${month}-${day}Z`, 'yyyy-M-dX', new Date());

        helpers.setValue(isValid(date) ? date : undefined);
      } else {
        helpers.setValue({ day, month, year });
      }

      setValues({
        ...values,
        [key]: inputValue,
      });
    },
    [helpers, type, values],
  );

  return (
    <FormFieldset
      id={id}
      legend={legend}
      legendSize={legendSize}
      hint={hint}
      error={errorMessage}
      useFormId={false}
      onBlur={async event => {
        event.persist();

        if (isFocused.current) {
          isFocused.current = false;
        }

        await delay();

        if (!isFocused.current) {
          field.onBlur(event);
        }
      }}
      onFocus={() => {
        isFocused.current = true;
      }}
    >
      {(partialDateType === 'dayMonthYear' || type === 'date') && (
        <FormGroup className="govuk-date-input__item">
          <FormNumberInput
            id={`${id}-day`}
            name={`${name as string}.day`}
            label="Day"
            width={2}
            value={parseNumber(values.day)}
            onChange={handleChange}
          />
        </FormGroup>
      )}

      <FormGroup className="govuk-date-input__item">
        <FormNumberInput
          id={`${id}-month`}
          name={`${name as string}.month`}
          label="Month"
          width={2}
          value={parseNumber(values.month)}
          onChange={handleChange}
        />
      </FormGroup>

      <FormGroup className="govuk-date-input__item">
        <FormNumberInput
          id={`${id}-year`}
          name={`${name as string}.year`}
          label="Year"
          width={4}
          value={parseNumber(values.year)}
          onChange={handleChange}
        />
      </FormGroup>
    </FormFieldset>
  );
}

export default FormFieldDateInput;
