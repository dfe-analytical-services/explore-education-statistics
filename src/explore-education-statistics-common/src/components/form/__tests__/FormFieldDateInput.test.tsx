import { Form } from '@common/components/form';
import FormFieldDateInput from '@common/components/form/FormFieldDateInput';
import { PartialDate } from '@common/utils/date/partialDate';
import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Formik } from 'formik';
import noop from 'lodash/noop';
import React from 'react';

describe('FormFieldDateInput', () => {
  test('renders correctly', () => {
    const { container } = render(
      <Formik initialValues={{}} onSubmit={noop}>
        <FormFieldDateInput
          legend="Start date"
          id="startDate"
          name="startDate"
        />
      </Formik>,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders with an error message correctly', () => {
    render(
      <Formik
        initialValues={{}}
        initialErrors={{
          startDate: 'The date is wrong',
        }}
        initialTouched={{
          startDate: true,
        }}
        onSubmit={noop}
      >
        <FormFieldDateInput
          legend="Start date"
          id="startDate"
          name="startDate"
        />
      </Formik>,
    );

    expect(screen.getByRole('group')).toHaveAttribute(
      'aria-describedby',
      'startDate-error',
    );
    expect(screen.getByText('The date is wrong')).toHaveAttribute(
      'id',
      'startDate-error',
    );
  });

  test('renders with a hint correctly', () => {
    render(
      <Formik initialValues={{}} onSubmit={noop}>
        <FormFieldDateInput
          legend="Start date"
          hint="Test hint"
          id="startDate"
          name="startDate"
        />
      </Formik>,
    );

    expect(screen.getByRole('group')).toHaveAttribute(
      'aria-describedby',
      'startDate-hint',
    );
    expect(screen.getByText('Test hint')).toHaveAttribute(
      'id',
      'startDate-hint',
    );
  });

  test('renders with a hint and error correctly', () => {
    render(
      <Formik
        initialValues={{}}
        initialErrors={{
          startDate: 'The date is wrong',
        }}
        initialTouched={{
          startDate: true,
        }}
        onSubmit={noop}
      >
        <FormFieldDateInput
          legend="Start date"
          hint="Test hint"
          id="startDate"
          name="startDate"
        />
      </Formik>,
    );

    expect(screen.getByRole('group')).toHaveAttribute(
      'aria-describedby',
      'startDate-error startDate-hint',
    );
    expect(screen.getByText('Test hint')).toHaveAttribute(
      'id',
      'startDate-hint',
    );
    expect(screen.getByText('The date is wrong')).toHaveAttribute(
      'id',
      'startDate-error',
    );
  });

  test('renders with correct defaults ids with form', () => {
    render(
      <Formik
        initialValues={{}}
        initialErrors={{
          startDate: 'The date is wrong',
        }}
        initialTouched={{
          startDate: true,
        }}
        onSubmit={noop}
      >
        <Form id="testForm">
          <FormFieldDateInput
            legend="Start date"
            hint="Test hint"
            name="startDate"
          />
        </Form>
      </Formik>,
    );

    const group = within(screen.getByRole('group'));

    expect(group.getByText('Test hint')).toHaveAttribute(
      'id',
      'testForm-startDate-hint',
    );
    expect(group.getByText('The date is wrong')).toHaveAttribute(
      'id',
      'testForm-startDate-error',
    );
    expect(group.getByLabelText('Day')).toHaveAttribute(
      'id',
      'testForm-startDate-day',
    );
    expect(group.getByLabelText('Month')).toHaveAttribute(
      'id',
      'testForm-startDate-month',
    );
    expect(group.getByLabelText('Year')).toHaveAttribute(
      'id',
      'testForm-startDate-year',
    );
  });

  test('renders with correct defaults ids without form', () => {
    render(
      <Formik
        initialValues={{}}
        initialErrors={{
          startDate: 'The date is wrong',
        }}
        initialTouched={{
          startDate: true,
        }}
        onSubmit={noop}
      >
        <FormFieldDateInput
          legend="Start date"
          hint="Test hint"
          name="startDate"
        />
      </Formik>,
    );

    const group = within(screen.getByRole('group'));

    expect(group.getByText('Test hint')).toHaveAttribute(
      'id',
      'startDate-hint',
    );
    expect(group.getByText('The date is wrong')).toHaveAttribute(
      'id',
      'startDate-error',
    );
    expect(group.getByLabelText('Day')).toHaveAttribute('id', 'startDate-day');
    expect(group.getByLabelText('Month')).toHaveAttribute(
      'id',
      'startDate-month',
    );
    expect(group.getByLabelText('Year')).toHaveAttribute(
      'id',
      'startDate-year',
    );
  });

  test('renders with correct custom ids with form', () => {
    render(
      <Formik
        initialValues={{}}
        initialErrors={{
          startDate: 'The date is wrong',
        }}
        initialTouched={{
          startDate: true,
        }}
        onSubmit={noop}
      >
        <Form id="testForm">
          <FormFieldDateInput
            legend="Start date"
            hint="Test hint"
            name="startDate"
            id="customId"
          />
        </Form>
      </Formik>,
    );

    const group = within(screen.getByRole('group'));

    expect(group.getByText('Test hint')).toHaveAttribute(
      'id',
      'testForm-customId-hint',
    );
    expect(group.getByText('The date is wrong')).toHaveAttribute(
      'id',
      'testForm-customId-error',
    );
    expect(group.getByLabelText('Day')).toHaveAttribute(
      'id',
      'testForm-customId-day',
    );
    expect(group.getByLabelText('Month')).toHaveAttribute(
      'id',
      'testForm-customId-month',
    );
    expect(group.getByLabelText('Year')).toHaveAttribute(
      'id',
      'testForm-customId-year',
    );
  });

  test('renders with correct custom ids without form', () => {
    render(
      <Formik
        initialValues={{}}
        initialErrors={{
          startDate: 'The date is wrong',
        }}
        initialTouched={{
          startDate: true,
        }}
        onSubmit={noop}
      >
        <FormFieldDateInput
          legend="Start date"
          hint="Test hint"
          name="startDate"
          id="customId"
        />
      </Formik>,
    );

    const group = within(screen.getByRole('group'));

    expect(group.getByText('Test hint')).toHaveAttribute('id', 'customId-hint');
    expect(group.getByText('The date is wrong')).toHaveAttribute(
      'id',
      'customId-error',
    );
    expect(group.getByLabelText('Day')).toHaveAttribute('id', 'customId-day');
    expect(group.getByLabelText('Month')).toHaveAttribute(
      'id',
      'customId-month',
    );
    expect(group.getByLabelText('Year')).toHaveAttribute('id', 'customId-year');
  });

  test('sets a valid UTC date as a form value', async () => {
    const onChange = jest.fn();

    render(
      <Formik<{ startDate?: Date }> initialValues={{}} onSubmit={noop}>
        {form => {
          onChange(form.values.startDate);

          return (
            <FormFieldDateInput
              legend="Start date"
              id="startDate"
              name="startDate"
            />
          );
        }}
      </Formik>,
    );

    userEvent.type(screen.getByLabelText('Day'), '10');
    userEvent.type(screen.getByLabelText('Month'), '12');
    userEvent.type(screen.getByLabelText('Year'), '2020');

    expect(screen.getByLabelText('Day')).toHaveValue(10);
    expect(screen.getByLabelText('Month')).toHaveValue(12);
    expect(screen.getByLabelText('Year')).toHaveValue(2020);

    expect(onChange).toHaveBeenCalledWith(new Date('2020-12-10T00:00:00.000Z'));
  });

  test('does not set an invalid date as a form value', async () => {
    const onChange = jest.fn();

    render(
      <Formik<{ startDate?: Date }> initialValues={{}} onSubmit={noop}>
        {form => {
          onChange(form.values.startDate);

          return (
            <FormFieldDateInput
              legend="Start date"
              id="startDate"
              name="startDate"
            />
          );
        }}
      </Formik>,
    );

    userEvent.type(screen.getByLabelText('Day'), '32');
    userEvent.type(screen.getByLabelText('Month'), '12');
    userEvent.type(screen.getByLabelText('Year'), '2020');

    expect(screen.getByLabelText('Day')).toHaveValue(32);
    expect(screen.getByLabelText('Month')).toHaveValue(12);
    expect(screen.getByLabelText('Year')).toHaveValue(2020);

    expect(onChange).toHaveBeenCalledWith(undefined);
  });

  test('does not set a partial date as a form value', async () => {
    const onChange = jest.fn();

    render(
      <Formik<{ startDate?: Date }> initialValues={{}} onSubmit={noop}>
        {form => {
          onChange(form.values.startDate);

          return (
            <FormFieldDateInput
              legend="Start date"
              id="startDate"
              name="startDate"
            />
          );
        }}
      </Formik>,
    );

    userEvent.type(screen.getByLabelText('Day'), '10');

    expect(screen.getByLabelText('Day')).toHaveValue(10);
    expect(onChange).toHaveBeenCalledWith(undefined);
  });

  test('can hide day field when `type = partialDate` and `partialDateType = monthYear`', () => {
    render(
      <Formik<{ startDate?: PartialDate }> initialValues={{}} onSubmit={noop}>
        <FormFieldDateInput
          legend="Start date"
          id="startDate"
          name="startDate"
          type="partialDate"
          partialDateType="monthYear"
        />
      </Formik>,
    );

    expect(screen.queryByLabelText('Day')).not.toBeInTheDocument();
    expect(screen.getByLabelText('Month')).toBeInTheDocument();
    expect(screen.getByLabelText('Year')).toBeInTheDocument();
  });

  test('does not hide day field when `type = date` and `partialDateType = monthYear`', () => {
    render(
      <Formik<{ startDate?: PartialDate }> initialValues={{}} onSubmit={noop}>
        <FormFieldDateInput
          legend="Start date"
          id="startDate"
          name="startDate"
          type="date"
          partialDateType="monthYear"
        />
      </Formik>,
    );

    expect(screen.getByLabelText('Day')).toBeInTheDocument();
    expect(screen.getByLabelText('Month')).toBeInTheDocument();
    expect(screen.getByLabelText('Year')).toBeInTheDocument();
  });

  test('can set a full PartialDate as a form value when `type = partialDate`', async () => {
    const onChange = jest.fn();

    render(
      <Formik<{ startDate?: PartialDate }> initialValues={{}} onSubmit={noop}>
        {form => {
          onChange(form.values.startDate);

          return (
            <FormFieldDateInput
              legend="Start date"
              id="startDate"
              name="startDate"
              type="partialDate"
            />
          );
        }}
      </Formik>,
    );

    userEvent.type(screen.getByLabelText('Day'), '15');
    userEvent.type(screen.getByLabelText('Month'), '6');
    userEvent.type(screen.getByLabelText('Year'), '2020');

    expect(screen.getByLabelText('Day')).toHaveValue(15);
    expect(screen.getByLabelText('Month')).toHaveValue(6);
    expect(screen.getByLabelText('Year')).toHaveValue(2020);

    expect(onChange).toHaveBeenCalledWith({
      day: 15,
      month: 6,
      year: 2020,
    });
  });

  test('can set only a day for a form value when `type = partialDate`', async () => {
    const onChange = jest.fn();

    render(
      <Formik<{ startDate?: PartialDate }> initialValues={{}} onSubmit={noop}>
        {form => {
          onChange(form.values.startDate);

          return (
            <FormFieldDateInput
              legend="Start date"
              id="startDate"
              name="startDate"
              type="partialDate"
            />
          );
        }}
      </Formik>,
    );

    userEvent.type(screen.getByLabelText('Day'), '15');

    expect(screen.getByLabelText('Day')).toHaveValue(15);

    expect(onChange).toHaveBeenCalledWith({
      day: 15,
    });
  });

  test('can set an invalid PartialDate as a form value when `type = partialDate`', async () => {
    const onChange = jest.fn();

    render(
      <Formik<{ startDate?: PartialDate }> initialValues={{}} onSubmit={noop}>
        {form => {
          onChange(form.values.startDate);

          return (
            <FormFieldDateInput
              legend="Start date"
              id="startDate"
              name="startDate"
              type="partialDate"
            />
          );
        }}
      </Formik>,
    );

    userEvent.type(screen.getByLabelText('Day'), '32');
    userEvent.type(screen.getByLabelText('Month'), '6');
    userEvent.type(screen.getByLabelText('Year'), '2020');

    expect(screen.getByLabelText('Day')).toHaveValue(32);
    expect(screen.getByLabelText('Month')).toHaveValue(6);
    expect(screen.getByLabelText('Year')).toHaveValue(2020);

    expect(onChange).toHaveBeenCalledWith({
      day: 32,
      month: 6,
      year: 2020,
    });
  });
});
