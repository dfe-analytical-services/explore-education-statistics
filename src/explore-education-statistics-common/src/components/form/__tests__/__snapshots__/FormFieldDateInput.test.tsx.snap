// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldDateInput renders correctly 1`] = `
<div class="govuk-form-group">
  <fieldset class="govuk-fieldset"
            id="startDate"
  >
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
      Start date
    </legend>
    <div class="govuk-form-group govuk-date-input__item">
      <label class="govuk-label"
             for="startDate-day"
      >
        Day
      </label>
      <input name="startDate.day"
             class="govuk-input govuk-input--width-2"
             id="startDate-day"
             type="number"
             value
      >
    </div>
    <div class="govuk-form-group govuk-date-input__item">
      <label class="govuk-label"
             for="startDate-month"
      >
        Month
      </label>
      <input name="startDate.month"
             class="govuk-input govuk-input--width-2"
             id="startDate-month"
             type="number"
             value
      >
    </div>
    <div class="govuk-form-group govuk-date-input__item">
      <label class="govuk-label"
             for="startDate-year"
      >
        Year
      </label>
      <input name="startDate.year"
             class="govuk-input govuk-input--width-4"
             id="startDate-year"
             type="number"
             value
      >
    </div>
  </fieldset>
</div>
`;
